### ISSUE TITLE
Automatic rewrites hang debugger

### ISSUE THREAD

**Issue by pavel-krivanek** [guillep/pharo-core#9356]
**Bug description**
When the code that you run in a debugger using Step Through calls indirectly a code with method rewrite (like deprecated methods), it hangs the debugger and if you manage to interrupt the frozen image, there is no obvious sign of what went wrong.

**To Reproduce**

Call a method 

```smalltalk
msg 

	self halt.
	self inside.
```

where the method inside is implemented like this: 
```smalltalk
inside

	self explore.
```
(`explore`is a deprecated method). 

Run the `msg` method and Step Through on halt.

Obvious partial solution is to turn off automatic rewrites during debugging.

**Comment by MarcusDenker** [comment]
Yes, we should turn it off. Do we have an easy way to check that we are running in the debugger?

**Comment by pavel-krivanek** [comment]
I do not know about any, maybe we should have a dedicated `ProcessLocalVariable` for that

**Comment by dionisiydk** [comment]
> it hangs the debugger and if you manage to interrupt the frozen image, there is no obvious sign of what went wrong.

Another case where a problem in the debugger implementation would not freeze the image if #8567 would be integrated. The UI process of the debugger would hang but cmd+. would recover.

**Comment by dionisiydk** [comment]
> Yes, we should turn it off. Do we have an easy way to check that we are running in the debugger?

It's too hardcore approach. It would not allow us to debug the deprecation rewrite engine.

**Comment by dionisiydk** [comment]
> I do not know about any, maybe we should have a dedicated `ProcessLocalVariable` for that

I would utilize CurrentExecutionEnvironment for this purpose. But the code with #ifUnderDebug: switches can be quite hard to debug.

**Comment by dionisiydk** [comment]
> > it hangs the debugger and if you manage to interrupt the frozen image, there is no obvious sign of what went wrong.
> 
> Another case where a problem in the debugger implementation would not freeze the image if #8567 would be integrated. The UI process of the debugger would hang but cmd+. would recover.

Hm, in my test it did not freeze the image. Only the debugger process. So cmd+. recovered the image. @pavel-krivanek did you have broken image at the end?

**Comment by pavel-krivanek** [comment]
It depends how long you let it hang. About 10 s make the image nonrecoverable
